source/common/upstream/upstream_impl.cc

```diff
@@ -834,80 +833,86 @@ bool BaseDynamicClusterImpl::updateDynamicHostList(const HostVector& new_hosts,
   // https://github.com/envoyproxy/envoy/issues/2874). We also check for duplicates here. It's
   // possible for DNS to return the same address multiple times, and a bad SDS implementation could
   // do the same thing.
-  std::unordered_set<std::string> host_addresses;
+
+  // Keep track of hosts we've seen during this pass so we can remove it from the
+  // list of current (i.e. existing hosts from this priority) hosts.
+  std::unordered_set<std::string> seen_hosts(current_hosts.size());
```

Comment:
This is related to the ongoing PR that @dio is working on where we are trying to make this code easier to understand. I actually think this PR is a huge improvement, but can you potentially either add more comments or think of different names for the variables? "seen_hosts" and "current_hosts" is still pretty confusing. In this code I would opt for very long and verbose names and/or lots of comments since it's so hard to understand.